### PR TITLE
Operational limits group

### LINKED ISSUE #2110: Operational limit sets
* **Do you want to request a *feature* or report a *bug*?**

A feature.

* **What is the current behavior?**

For a line or a transformer, we have only one group (permanent limit + temporary limits for two sides) of each limit type (current, active power and apparent power) available in the core model, such as:
````
<currentLimits1 permanentLimit="580"/>
<currentLimits2 permanentLimit="580">
      <temporaryLimit name="N/A" value="734"/>
      <temporaryLimit name="IT105" acceptableDuration="6300" value="580"/>
      <temporaryLimit name="IT1" acceptableDuration="60"/>
</currentLimits2>
````
By definition, as we have only one group, it means that it is the active one. By active, we mean the one that has to be considered by the analysis tools (security analysis for example). 

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

In some studies (network development) or operational context (CGMES), we can have a set of operational limits. The season can be important to consider, the day vs the night, and many other factor that can lead to store a set of limits and not one, and let the user chooses the active set. 
- Important : I think that for reading and writing performance issue, we must add a parameter that forces import and export of only active limits (CGMES, internal importers). 
- We must have three `OperationalLimitSet`, one for current, one for active power and one for apparent power. The design looks nice in Miora's PR (see below), excepted that `activated` must be changed for `enabled` to define the active limit set. 
- Inside a set, each group (permanent limit + temporary limit for 2 sides) must have an id (String).
- In a set, the active group of limits is designed through its id (String).
- Most of the use cases, no breaking changes must appear. 
- Maybe, High or low for the modelling interpretation of the whole set. Note in that PR, to be checked with CGMES modelling.

This leads to a new version of IIDM (1.12).

Note that we must think about additional information in `CurrentLimits`, but not in that PR:
- A type can be added to a group. This type will described if the group is linked to an automaton/actions, or if it is only for supervision.
- An enum to distinguish protections from study limits.

* **What is the expected behavior?**

* **What is the motivation / use case for changing the behavior?**

* **Please tell us about your environment:**
  - PowSyBl Version: ...
  - OS Version: ...

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, spectrum, etc)

(if a question doesn't apply, you can delete it)

### PR DESCRIPTION
**Pull request by pjeanmarie**
**Please check if the PR fulfills these requirements**

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

Fixes #2110

**Does this PR introduce a breaking change or deprecate an API?**
- [X] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [X] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

If you have defined your own IIDM implementation, you should:
- implement the `OperationalLimitGroup` interface;

- in you `Branch` implementations:
  - implement the following methods:
  ```java
    Collection<OperationalLimitsGroup> getOperationalLimitsGroups1();
    Optional<String> getSelectedOperationalLimitsGroupId1();
    Optional<OperationalLimitsGroup> getOperationalLimitsGroup1(String id);
    Optional<OperationalLimitsGroup> getSelectedOperationalLimitsGroup1();
    OperationalLimitsGroup newOperationalLimitsGroup1(String id);
    void setSelectedOperationalLimitsGroup1(String id);
    void removeOperationalLimitsGroup1(String id);
    void cancelSelectedOperationalLimitsGroup1();

    Collection<OperationalLimitsGroup> getOperationalLimitsGroups2();
    Optional<String> getSelectedOperationalLimitsGroupId2();
    Optional<OperationalLimitsGroup> getOperationalLimitsGroup2(String id);
    Optional<OperationalLimitsGroup> getSelectedOperationalLimitsGroup2();
    OperationalLimitsGroup newOperationalLimitsGroup2(String id);
    void setSelectedOperationalLimitsGroup2(String id);
    void removeOperationalLimitsGroup2(String id);
    void cancelSelectedOperationalLimitsGroup2();
  ```
  - redefine your implementation of the following methods:

  ```java
    Optional<CurrentLimits> getCurrentLimits1();
    CurrentLimits getNullableCurrentLimits1();
    Optional<ActivePowerLimits> getActivePowerLimits1();
    ActivePowerLimits getNullableActivePowerLimits1();
    Optional<ApparentPowerLimits> getApparentPowerLimits1();
    ApparentPowerLimits getNullableApparentPowerLimits1();
    CurrentLimitsAdder newCurrentLimits1();
    ActivePowerLimitsAdder newActivePowerLimits1();
    ApparentPowerLimitsAdder newApparentPowerLimits1();

    Optional<CurrentLimits> getCurrentLimits2();
    CurrentLimits getNullableCurrentLimits2();
    Optional<ActivePowerLimits> getActivePowerLimits2();
    ActivePowerLimits getNullableActivePowerLimits2();
    Optional<ApparentPowerLimits> getApparentPowerLimits2();
    ApparentPowerLimits getNullableApparentPowerLimits2();
    CurrentLimitsAdder newCurrentLimits2();
    ActivePowerLimitsAdder newActivePowerLimits2();
    ApparentPowerLimitsAdder newApparentPowerLimits2();
  ```
...

**Other information**:

### CODE REVIEW

**Comment by pjeanmarie** [comment]
Au niveau de l'adder, pour info:
newLimits(<id>) de newCurrentLimitsSet1() est équivalent en fonctionnement au beginTemporaryLimit() et addToSet() est équivalent à  endTemporaryLimit()
bien sûr les noms (entre autre) peuvent être changés

File: iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/AbstractCurrentLimitsTest.java [hunk 1]
```diff
@@ -196,6 +196,93 @@ private static Network createNetworkForThreeWindingsTransformer() {
...
+                    .setName("1'")
+                    .setAcceptableDuration(60)
+                    .setValue(1600.0)
+                    .endTemporaryLimit()
+                .addToSet()
+                .newLimits("3")
+                    .setPermanentLimit(800.0)
+                .addToSet()
+                .add();
+        l.newCurrentLimitsSet2().newLimits("1").setPermanentLimit(700.0).addToSet().enable("1").add();
```
[the remark this answers is not in the document]
Comment (pjeanmarie) [reply iidm/iidm-tck/src/test/java/com/powsybl/iidm/network/tck/AbstractCurrentLimitsTest.java:281]: newLimits(<id>) in an OperationalLimitsSet will be added in addition to the newLimits(<id>) in the OperationalLimitsSetAdder for this use case

**Comment by pjeanmarie** [comment]
@geofjamg @annetill Je vais changer l'héritage dans la mesure où LoadingLimits et son Adder ne vont plus étendre OperationalLimits et son Adder mais c'est le Set qui va les étendre, c'est ok ou ça casse qqch d'important?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/OperationalLimitsGroups.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import com.powsybl.iidm.network.OperationalLimitsGroup;
+import com.powsybl.iidm.network.Validable;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * @author Miora Ralambotiana {@literal <miora.ralambotiana at rte-france.com>}
+ */
+public interface OperationalLimitsGroups {
```
Comment (annetill) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/OperationalLimitsGroups.java:21]: Should be in the API, no?

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/FlowsLimitsDefaultHolder1.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * SPDX-License-Identifier: MPL-2.0
+ */
+package com.powsybl.iidm.network.impl;
+
+import com.powsybl.iidm.network.*;
+
+import java.util.List;
+import java.util.Optional;
+
+interface FlowsLimitsDefaultHolder1 extends FlowsLimitsHolder1 {
```
Comment (flo-dup) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/FlowsLimitsDefaultHolder1.java:15]: This should be an abstract class, not an interface

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/ActivePowerLimitsAndGroupAdderImpl.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * SPDX-License-Identifier: MPL-2.0
+ */
+package com.powsybl.iidm.network.impl;
+
+import com.powsybl.iidm.network.*;
+
+import java.util.Optional;
+
+/**
+ */
```
Comment (flo-dup) [inline iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/ActivePowerLimitsAndGroupAdderImpl.java:15]: authorship is missing on the new files

**Review by flo-dup — DISMISSED** [review]

**Review by olperr1 — CHANGES_REQUESTED** [review]

File: iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/OperationalLimitsGroupsImpl.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                .or(() -> Optional.ofNullable(operationalLimitsGroupById.get(DEFAULT_OPERATIONAL_LIMITS_GROUP_DEFAULT_ID)))
+                .orElseGet(() -> newOperationalLimitsGroup(DEFAULT_OPERATIONAL_LIMITS_GROUP_DEFAULT_ID));
+    }
+
+    @Override
+    public CurrentLimitsAdder newCurrentLimits() {
+        OperationalLimitsGroupImpl group = getOrCreateDefaultOperationalLimitsGroup();
+        setDefaultOperationalLimitsGroup(group.getId());
+        return new CurrentLimitsAdderImpl(getOrCreateDefaultOperationalLimitsGroup(), identifiable, identifiable.getId());
+    }
```
[the remark this answers is not in the document]
Comment (olperr1) [reply iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/OperationalLimitsGroupsImpl.java]: Fixed

**Pull request merged**